synapse/http/proxyagent.py

```diff
@@ -208,30 +210,48 @@ def request(self, method, uri, headers=None, bodyProducer=None):
         )
 
 
-def _http_proxy_endpoint(proxy, reactor, **kwargs):
+def _http_proxy_endpoint(
+    proxy: Optional[bytes],
+    reactor: IReactorCore,
+    tls_options_factory: Optional[IPolicyForHTTPS],
+    **kwargs
+):
     """Parses an http proxy setting and returns an endpoint for the proxy
+    This currently supports http:// and https:// proxies. a hostname without scheme is assumed to be http.
 
     Args:
         proxy (bytes|None):  the proxy setting
         reactor: reactor to be used to connect to the proxy
+        tls_options_factory: the TLS options to use when connecting through a https proxy
         kwargs: other args to be passed to HostnameEndpoint
 
     Returns:
         interfaces.IStreamClientEndpoint|None: endpoint to use to connect to the proxy,
             or None
+
+    Raises:
+        ValueError if given a proxy with an unsupported scheme.
     """
     if proxy is None:
         return None
 
-    # currently we only support hostname:port. Some apps also support
-    # protocol://<host>[:port], which allows a way of requiring a TLS connection to the
-    # proxy.
+    # Note: we can't use urlsplit/urlparse as that is completely broken for things without a scheme://
+    schemehost, port = parse_host_port(proxy, default_port=1080)
+    scheme, host = parse_scheme_host(schemehost, default_scheme=b"http")
```

Comment:
I think it makes more sense to rename this into `parse_proxy` which returns a `scheme, host, port`. I'd also like to see us add tests for it while we're here! (They should be very easy to write!) In particular we should test:

* Host-only
* Host + port
* Scheme + host
* Scheme + host + port

The above should also get tested when host is an IPv4 address and when it is an IPv6 address.

I'm a bit worried that this is going to be a bit fiddly, note that the tests in https://github.com/python/cpython/blob/3.9/Lib/test/test_urlparse.py could be useful to look over.

I do wonder if we should use urlsplit and then "fix" it if there's not a scheme...that seems a bit fiddly across Python versions though.

Reply:
> Rename this into parse_proxy

I guess so. will do!

The tests I added to this MR *do* test exactly this, I believe: https://github.com/matrix-org/synapse/pull/9119/files#diff-65bc4691f51cf0a8187bead3590b97e3ade05004a15e90ae6f40410d77fd3eeaR456

It's missing a few cases though, namely:

* no port
* ipv4/6 addresses

Should I just add those to the tests I wrote, or should the tests happen at a different place?


> I do wonder if we should use urlsplit and then "fix" it if there's not a scheme...that seems a bit fiddly across Python versions though.

I think I tried going that route and not liking where I ended up.

Reply:
I disagree that they test this. I'm asking to test it a much lower-level, e.g. directly call this function with different strings and see that it returns what we want.

Reply:
I'm reworking this currently. One thing I'm not sure about is how to handle ipv6 literals. When part of a url specification they are expected to be wrapped into `[]` because otherwise it might lead to ambiguities with the port specification.

But I guess synapse internally needs the raw ipv6 ip without the enclosing brackets? I'll try to add hat to the method, I just wanted to confirm that I'm on the right track here.

Reply:
Mh, there's also something called a zone ID that's sometimes part of ipv6 adresses: https://veithen.io/2013/12/30/how-to-correctly-parse-ipv6-addresses.html

This is further complicated by the fact that for uri's you are expected to urlencode this, so the following might be something that's passed in an env var to synapse:
`https://`

I have no idea what this is or how synapse/twisted would handle  that internally, if at all. :-/

Reply:
ipv6 problems (which exist in the current code as well) moved to a later PR.

Reply:
I've moved this to a new `parse_proxy` method and added a boatload of tests for it :).

Some of the ipv6 specific test cases have commented out stuff that's currently not handled correctly which needs to be fixed separately, as discussed in #synapse-dev.